File: GESDISC/MERRA300.prod.assim.inst3_3d_chm_Ne.20021201.hdf.py
"""
Copyright (C) 2020 The HDF Group

This example code illustrates how to access a GESDISC MERRA file in Python
and save it as an netCDF-4 file. Then, it reads it back to write an
equivalent HDF4 file.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python MERRA300.prod.assim.inst3_3d_chm_Ne.20021201.hdf.py

The HDF file must be in your current working directory.

Tested under: Python 3.7.3 :: Anaconda
Last updated: 2020-4-06
"""
import os
import numpy as np
from netCDF4 import Dataset
from pyhdf.SD import SD, SDC

def write_netcdf4(_fname, _lat, _lon, _data, _varname, _long_name, _units):
    # Open a netCDF file to write.
    ncout = Dataset(_fname, 'w', format='NETCDF4')
    
    # Define axis size.
    nlat = len(_lat)
    nlon = len(_lon)    
    ncout.createDimension('lat', nlat)
    ncout.createDimension('lon', nlon)
    
    # Create latitude axis.
    lat = ncout.createVariable('lat', np.dtype('double').char, ('lat'))
    lat.standard_name = 'latitude'
    lat.long_name = 'latitude'
    lat.units = 'degrees_north'
    lat.axis = 'Y'
    
    # Create longitude axis.
    lon = ncout.createVariable('lon', np.dtype('double').char, ('lon'))
    lon.standard_name = 'longitude'
    lon.long_name = 'longitude'
    lon.units = 'degrees_east'
    lon.axis = 'X'

    # Create variable array.
    vout = ncout.createVariable(_varname, np.dtype('double').char,
                                ('lat', 'lon'))
    vout.long_name = _long_name
    vout.units = _units
    # Copy axis from original dataset.
    lon[:] = _lon[:]
    lat[:] = _lat[:]
    vout[:] = _data[:]

    # Close file.
    ncout.close()

def write_hdf(_fname, _lat, _lon, _data, _var_name, _long_name, _units):
    # Create file.
    d = SD(_fname, SDC.WRITE|SDC.CREATE) 

    # Create lat.
    nlat = len(_lat)
    lat = d.create('lat', SDC.FLOAT64, nlat)
    d0 = lat.dim(0)
    d0.setname('YDim:EOSGRID')
    lat[:] = _lat
    setattr(lat, 'units', 'degrees_north')
    
    # Create lon
    nlon = len(_lon)    
    lon = d.create('lon', SDC.FLOAT64, nlon)
    d1 = lon.dim(0)    
    d1.setname('XDim:EOSGRID')
    lon[:] = _lon
    setattr(lon, 'units', 'degrees_east')
    
    # Create var.
    v = d.create(_var_name, SDC.FLOAT64, (nlat, nlon))
    v0 = v.dim(0)
    v1 = v.dim(1)
    v0.setname('YDim:EOSGRID')
    v1.setname('XDim:EOSGRID')
    v[:] = _data    
    setattr(v, 'long_name', _long_name)
    setattr(v, 'units', _units)

    # Close datasets.
    v.endaccess()
    lon.endaccess()
    lat.endaccess()
    d.end()
    
    
    
FILE_NAME = 'MERRA300.prod.assim.inst3_3d_chm_Ne.20021201.hdf'
DATAFIELD_NAME = 'PLE'
    
# Read HDF4 file.
hdf = SD(FILE_NAME, SDC.READ)

# Read dataset.
data4D = hdf.select(DATAFIELD_NAME)

# Subset data.
data = data4D[0,72,:,:].astype(np.float64)

# Retrieve the attributes.
attrs = data4D.attributes(full=1)
mva=attrs["missing_value"]
missing_value = mva[0]
lna=attrs["long_name"]
long_name = lna[0]
ua=attrs["units"]
units = ua[0]        

# Read geolocation dataset.
lat = hdf.select('YDim')
latitude = lat[:]
lon = hdf.select('XDim')
longitude = lon[:]

# Replace the missing values with NaN.        
data[data == missing_value] = np.nan
datam = np.ma.masked_array(data, np.isnan(data))

# Close datasets.
lat.endaccess()
lon.endaccess()
data4D.endaccess()

# Close file.
hdf.end()

# Write netCDF-4 file.
write_netcdf4(FILE_NAME+'.nc4', latitude, longitude, datam, DATAFIELD_NAME,
              long_name, units)
             
# Read netCDF-4 file.
nc = Dataset(FILE_NAME+'.nc4')
data = nc.variables[DATAFIELD_NAME][:, :].astype(np.float64)
    
# Retrieve the attributes.
# missing_value = nc.variables[DATAFIELD_NAME].missing_value
long_name = nc.variables[DATAFIELD_NAME].long_name
units = nc.variables[DATAFIELD_NAME].units

# Retrieve the geolocation data.
latitude = nc.variables['lat'][:]
longitude = nc.variables['lon'][:]

# Write HDF4 file.
write_hdf(FILE_NAME+'.nc4.hdf', latitude, longitude, data, DATAFIELD_NAME,
          long_name, units)

    
    

File: GESDISC/MERRA300_prod_assim_inst3_3d_chm_Ne_20021201_hdf.m
%
%  This example code illustrates how to access and visualize
% GES DISC MERRA HDF-EOS2 Grid file in MATLAB. 
%
%  If you have any questions, suggestions, comments  on this example, 
% pleaseuse the HDF-EOS Forum (http://hdfeos.org/forums). 
%
%  If you would like to see an  example of any other NASA
% HDF/HDF-EOS data product that is not listed in the
% HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), 
% feel free to contact us at eoshelp@hdfgroup.org or post it at the
% HDF-EOS Forum (http://hdfeos.org/forums).
%
% Usage:save this script and run (without .m at the end)
%
% $matlab -nosplash -nodesktop -r MERRA300_prod_assim_inst3_3d_chm_Ne_20021201_hdf
%
% Tested under: MATLAB R2020a
% Last updated: 2020-10-21

import matlab.io.hdf4.*
import matlab.io.hdfeos.*

% Open the HDF-EOS2 Grid file.
file_name='MERRA300.prod.assim.inst3_3d_chm_Ne.20021201.hdf';
file_id = gd.open(file_name, 'rdonly');

% Read data from a data field.
grid_name='EOSGRID';
grid_id = gd.attach(file_id, grid_name);

datafield_name='PLE';

[data1, lat, lon] = gd.readField(grid_id, datafield_name, [], [], []);

% Convert 4-D data to 2-D data.
data=squeeze(data1(:,:,73,2));

% Convert the data to double type for plot.
data=double(data);


% Detach from the grid object.
gd.detach(grid_id);

% Close the file.
gd.close(file_id);

% Read attributes from the data field.
SD_id = sd.start(file_name, 'rdonly');
sds_index = sd.nameToIndex(SD_id, datafield_name);
sds_id = sd.select(SD_id, sds_index);

% Read fillvalue and missing value from the data field.
fillvalue_index = sd.findAttr(sds_id, '_FillValue');
fillvalue = sd.readAttr(sds_id, fillvalue_index);
missingvalue_index = sd.findAttr(sds_id, 'missing_value');
missingvalue = sd.readAttr(sds_id, missingvalue_index);

% Read long_name from the data field.
long_name_index = sd.findAttr(sds_id, 'long_name');
long_name = sd.readAttr(sds_id, long_name_index);

% Read units from the data field.
units_index = sd.findAttr(sds_id, 'units');
units = sd.readAttr(sds_id, units_index);

% Read scale_factor from a data field.
scale_index = sd.findAttr(sds_id, 'scale_factor');
scale = sd.readAttr(sds_id, scale_index);

% Convert to double type for plot.
scale = double(scale);

% Read add_offset from a data field.
offset_index = sd.findAttr(sds_id, 'add_offset');
offset = sd.readAttr(sds_id, offset_index);

% Convert to double type for plot.
offset = double(offset);

% Terminate access to the corresponding data set.
sd.endAccess(sds_id);

% Close the file.
sd.close(SD_id);

% Replace the filled value with NaN.
data(data==fillvalue) = NaN;

% Replace the missing value with NaN.
data(data==missingvalue) = NaN;

% Multiply scale and adding offset.
data = data*scale + offset ;

% Plot the data.
latlim = [floor(min(min(lat))),ceil(max(max(lat)))];
lonlim = [floor(min(min(lon))),ceil(max(max(lon)))];
min_data = floor(min(min(data)));
max_data = ceil(max(max(data)));

f = figure('Name', file_name, 'visible', 'off');

axesm('MapProjection','eqdcylin','Frame','on','Grid','on', ...
      'MapLatLimit',latlim,'MapLonLimit',lonlim, ...
      'MeridianLabel','on','ParallelLabel','on','MLabelParallel','south');
coast = load('coast.mat');
lat = fliplr(lat);
surfm(lat,lon,data);
colormap('Jet');
caxis([min_data max_data]); 

% Change the value if you want to have more than 10 tick marks.
ntickmarks = 10;
granule = floor((max_data - min_data) / ntickmarks);

h=colorbar('YTick', min_data:granule:max_data);

plotm(coast.lat,coast.long,'k')

title({file_name; ...
       [long_name  [' at TIME=1 and Height=72']]}, ...
       'Interpreter', 'None', 'FontSize', 16,'FontWeight','bold');
tightmap;
set(get(h, 'title'), 'string', units, 'FontSize', 8,'FontWeight','bold');
saveas(f, [file_name '.m.png']);
exit;
